pkg/csi/service/common/authmanager.go

```diff
@@ -135,44 +184,237 @@ func GenerateDatastoreIgnoreMapForBlockVolumes(ctx context.Context, vc *vsphere.
 		for dsURL, dsInfo := range dsURLTodsInfoMap {
 			dsURLs = append(dsURLs, dsURL)
 			dsInfos = append(dsInfos, dsInfo)
+			dsMoRef := dsInfo.Reference()
+			entities = append(entities, dsMoRef)
+		}
+	}
+
+	dsURLToInfoMap, err := getDatastoresWithRequiredPriv(ctx, vc, dsURLs, dsInfos, entities)
+	if err != nil {
+		log.Errorf("failed to get datastores with required priv. Error: %+v", err)
+		return nil, err
+	}
+	return dsURLToInfoMap, nil
+}
+
+// GenerateDatastoreMapForFileVolumes scans all datastores in Vcenter and do privilege check on those datastoes
+// It will return datastores which has the privileges for creating file volume
+func GenerateDatastoreMapForFileVolumes(ctx context.Context, vc *vsphere.VirtualCenter) (map[string]*cnsvsphere.DatastoreInfo, error) {
+	log := logger.GetLogger(ctx)
+	// get all vSAN datastores from VC
+	vsanDsURLToInfoMap, err := vc.GetVsanDatastores(ctx)
+	if err != nil {
+		log.Errorf("failed to get vSAN datastores with error %+v", err)
+		return nil, err
+	}
+	var allvsanDatastoreUrls []string
+	for dsURL := range vsanDsURLToInfoMap {
+		allvsanDatastoreUrls = append(allvsanDatastoreUrls, dsURL)
+	}
+	fsEnabledMap, err := IsFileServiceEnabled(ctx, allvsanDatastoreUrls, vc)
+	if err != nil {
+		log.Errorf("failed to get if file service is enabled on vsan datastores with error %+v", err)
+		return nil, err
+	}
+
+	var dsURLs []string
+	var dsInfos []*cnsvsphere.DatastoreInfo
+	var entities []types.ManagedObjectReference
+	dsURLToInfoMap := make(map[string]*cnsvsphere.DatastoreInfo)
+	for dsURL, dsInfo := range vsanDsURLToInfoMap {
+		if val, ok := fsEnabledMap[dsURL]; ok {
+			if val {
+				// go through all datastores to build entities
+				dsMoRef := dsInfo.Reference()
+				entities = append(entities, dsMoRef)
+				// Get all vSAN datastore URLs and Info which has vSAN FS enabled
+				dsURLs = append(dsURLs, dsURL)
+				dsInfos = append(dsInfos, dsInfo)
+				dsURLToInfoMap[dsURL] = dsInfo
+			}
 		}
 	}
+	log.Debugf("dsURLToInfoMap is %+v", dsURLToInfoMap)
+	return dsURLToInfoMap, nil
+}
 
-	log.Debugf("auth manager: dsURLs %v dsInfos %v", dsURLs, dsInfos)
-	dsIgnoreURLToInfoMap := make(map[string]*cnsvsphere.DatastoreInfo)
+// IsFileServiceEnabled checks if file service is enabled on the specified datastoreUrls.
+func IsFileServiceEnabled(ctx context.Context, datastoreUrls []string, vc *vsphere.VirtualCenter) (map[string]bool, error) {
+	// Compute this map during controller init. Re use the map every other time.
+	log := logger.GetLogger(ctx)
+	err := vc.Connect(ctx)
+	if err != nil {
+		log.Errorf("failed to connect to VirtualCenter. err=%v", err)
+		return nil, err
+	}
+	err = vc.ConnectVsan(ctx)
+	if err != nil {
+		log.Errorf("error occurred while connecting to VSAN, err: %+v", err)
+		return nil, err
+	}
+	// This gets the datastore to file service enabled map for all the vsan datastores.
+	dsToFileServiceEnabledMap, err := getDsToFileServiceEnabledMap(ctx, vc)
```

Comment:
We seem to be listing datacenters and getting datastores from the datacenters 2 times - once in GenerateDatastoreMapForFileVolumes and once in getDsToFileServiceEnabledMap. Can we make the VC API calls just once?

Reply:
This comment does not seem to be resolved.